Transparent ascription of modules

mrmr1993 opened the pull request:
This PR adds support for 'transparent ascription' of modules.

This takes the syntax:
```ocaml
module M : sig
  type a
  type b
  val x : a
  val y : b
end

(* signatures *)
module type Restrict_M = sig
  module N = (M :> sig type a val x : a end)
end

(* structures *)
module N = (M :> sig type a val x : a end)
```

This syntax declares that `N` is 'the same' as `M` -- in the sense that any of its members will resolve to members of `M` -- but that its accessible members are the subset exposed by the signature. Concretely, in the example above:
* `N.a = M.a`
* `N.x = M.x`
* `N.b` and `N.y` do not exist
* For any functor `F` that accepts a `sig type a val x : a end`, `F(N) = F(M)`
* For any module type `S` a subtype of `sig type a val x : a end`, `(N :> S) = (M :> S)`

See `testsuite/tests/typing-modules/transparent_ascription.ml` for tests of these and other behaviours.

The most interesting uses of this feature are in aliases to functor arguments (e.g. now `functor (A : S) -> struct module A = A end` can track that `F(M).A = (M :> S)`), and in a cleaner syntax for `include`s/`open`s (e.g. `include (M :> S)` to include only some desired members of `M`). The first of these means that we infer stronger signatures for some functors, which changes the expected output of several tests; the reported types of some user code will change similarly.

There was some previous discussion about this feature in #10435.

---------------------

Implementation:

This adds an optional `module_type` argument to `Mty_alias`, where `Mty_alias ("M", None)` represents `= M`, and `Mty_alias ("M", Some "S")` represents `= (M :> S)`.

The most significant changes are those in `typing/typemod.ml` and `typing/includemod.ml`.

mrmr1993 committed:
Add syntax for transparent ascription in module types

mrmr1993 committed:
Add Mty_ascribe module type

mrmr1993 committed:
Propagate ascriptions when strengthening

mrmr1993 committed:
Fixup compilation of ocamldoc

mrmr1993 committed:
Relax constraints on aliases with ascribed types

mrmr1993 committed:
Use transparent ascription for aliases to functor arguments

mrmr1993 committed:
Remove ascriptions in functor arguments

mrmr1993 committed:
Normalise paths through type ascriptions where safe to do so

mrmr1993 committed:
Don't use ascriptions for existing ascribed or aliased module types

mrmr1993 committed:
Fix typo in signature module strengthening (ascribe -> alias)

mrmr1993 committed:
Weaken aliases to ascriptions for functor args in the environment

mrmr1993 committed:
Fix behaviour of aliases interacting with ascriptions

mrmr1993 committed:
Update expected test outputs, add an additional related test for pr7818

mrmr1993 committed:
Add tests for transparent ascription

mrmr1993 committed:
Add module syntax for constructing ascriptions

mrmr1993 committed:
Ascription syntax tests

mrmr1993 committed:
Fix ocamldoc compilation

mrmr1993 committed:
Merge Mty_alias and Mty_ascribe

mrmr1993 committed:
Bootstrap

mrmr1993 committed:
Typos

mrmr1993 committed:
Remove completed TODO, tidy up superfluous formatting changes

mrmr1993 committed:
Move Translmod.compose_coercions to Includemod

mrmr1993 committed:
Fix ascription for paths with existing implicit coercions

mrmr1993 committed:
Add tests for using ascripted modules in `include`s and expressions

lpw25 commented:
I've had an initial look through. This looks good. My main issue with it is that:
```ocaml
module Foo (X : sig val x : int end) = struct
  module M = X
end
```
is currently given type:
```ocaml
module Foo :
  functor (X : sig val x : int end) ->
    sig module M = (X :> sig val x : int end) end
```
when it would be better to give it the type:
```ocaml
module Foo :
  functor (X : sig val x : int end) ->
    sig module M = X end
```
and then have functor application deal with the transparent ascription. As mentioned in [my comment on #10435](https://github.com/ocaml/ocaml/pull/10435#issuecomment-853791493), this approach is more robust as it entirely removes the need to track the unstable property of being a functor parameter. There is a lot of complexity in the current system around trying (and failing) to track what is a functor parameter (e.g. `Env.is_functor_arg`, `Mtype.scrape_for_functor_arg`), all of which can be removed if we use the alternative approach.

Another issue is that I think we need to add aliases to functor applications as a prerequisite to this. Since functor arguments can be functor applications, this patch already allows aliases to functor applications in some places, e.g.:
```ocaml
# module Baz = struct let y = 43 end;;
module Baz : sig val y : int end
# module Bar (X : sig val y : int end) = struct let x = X.y + 1 end;;
module Bar : functor (X : sig val y : int end) -> sig val x : int end
# let foo () =
    let module M = Foo(Bar(Baz)) in
    M.M.x;;
val foo : unit -> int = <fun>
```
but complains about them in others:
```ocaml
# module M = Foo(Bar(Baz));;
Line 1:
Error: In module M:
       Modules do not match:
         sig module M = M.M end
       is not included in
         sig module M = (Bar(Baz) :> sig val x : int end) end
       In module M.M: Module Bar(Baz) cannot be aliased
```

mrmr1993 committed:
Merge remote-tracking branch 'upstream/trunk' into feature/transparent-ascription2

mrmr1993 committed:
Bootstrap

mrmr1993 committed:
Update .depend

mrmr1993 committed:
Update changelog

mrmr1993 committed:
Allow the construction of ascribed functors

mrmr1993 committed:
Add support for extended paths in functor descriptions

mrmr1993 committed:
Add (X :> _) syntax for 'ascribe to underlying type'.

mrmr1993 committed:
Update expected test output

mrmr1993 committed:
Add tests for ascription with an implicit type

mrmr1993 committed:
Allow ascription of modules other than paths

mrmr1993 committed:
Add tests for ascription of arbitrary module expressions

mrmr1993 committed:
Merge remote-tracking branch 'upstream/trunk' into feature/transparent-ascription2

mrmr1993 committed:
Bootstrap

mrmr1993 commented:
I've updated this, adding support for ascriptions of functors (`module M = (F(A) :> S)` in signatures and structures) to avoid the inconsistency above.

> I've had an initial look through. This looks good. My main issue with it is that:
> 
> ```ocaml
> module Foo (X : sig val x : int end) = struct
>   module M = X
> end
> ```
> 
> is currently given type:
> 
> ```ocaml
> module Foo :
>   functor (X : sig val x : int end) ->
>     sig module M = (X :> sig val x : int end) end
> ```
> 
> when it would be better to give it the type:
> 
> ```ocaml
> module Foo :
>   functor (X : sig val x : int end) ->
>     sig module M = X end
> ```
> 
> and then have functor application deal with the transparent ascription

I've been trying to avoid making all modules `Mp_present`, and without that we may have to coerce the result of each functor application immediately to make these aliases present. There's also the added difficulty of looking up the module type for the signature from within `Subst.modtype` and friends. Instead, I've added a syntactic hack, allowing an implicit ascription
```ocaml
module M' = (M :> _)
```
which ascribes the module's own type to it. Using this for functors, we avoid spelling out the module type of the ascription until it becomes relevant after application.
```ocaml
# module F(M : sig type t end) = struct
  module N = M
end;;
module F : functor (M : sig type t end) -> sig module N = (M :> _) end
# module G = struct type t = int let next : t -> t = (+) 1 end;;
module G : sig type t = int val next : t -> t end
# module A = F(G);;
module A : sig module N = (G :> sig type t end) end
```
This also 'allows' functor aliases without dealing with the complexities of presence for normal functor aliases:
```ocaml
# module B = (F(G) :> _);;
module B = (F(G) :> _)
```

lpw25 commented:
> I've been trying to avoid making all modules Mp_present

I've also spent a lot of time trying to find systems that keep `Mp_absent` as the default. However, I've come to the conclusion that all such systems are complex and fragile, and that the better approach is to bite the bullet and just make `Mp_present` the default.

The only place where people actually need `Mp_absent` is when using aliases for namespacing as e.g. dune does. To handle such cases it is sufficient to add a `no_alias_deps` parameter to the `Typemod.type_module` and friends that is initially set to the value of `Clflags.no_alias_deps`. Then only create `Mp_absent` aliases when that parameter is set.

This approach ensures that `Mp_absent` never appears in a named module type or a functor body, and so it's not possible to unexpectedly get an `Mp_absent` alias to a functor application or a module ascription.

mrmr1993 commented:
I would still prefer to go for a softer approach. As far as I can see, there are 3 places where we need to be careful:
* module types, where functor aliases or rewriting may equate an absent alias to a present one
```ocaml
module type S = sig
  module N = F(A)
end
module type S' = sig
  module M : sig end
  module N = M
end
module type S'' = S' with module M := F(A)
module X : sig module type S = S end = struct
  module type S = S''
end
```
* packed modules, where presence/absence is vitally important
* reordering of module aliases within signatures

With this in mind, I've been playing with an approach that makes
* all named module types have `Mp_present` aliases
  - this prevents module reordering at inclusion-checking time
* all 'normal' modules in the toplevel structure/signature (and also in `let module M = ... in`) have `Mp_absent` aliases
* one of the above modules constrained by a named module type (`module M : S = ...`) use a variant of the named type with the aliases made absent
* packed modules use `Mp_present` for all aliases, matching the named signature
* functor arguments and results use `Mp_present` for all modules
  - I suspect this is unnecessary; I'll speak more to this below

This seems to work well (modulo a `module type of` bug I'm trying to track down), and avoids eagerly linking external modules and all of their symbols for most aliases, as well as avoiding issues with the dune-style toplevel alias module.

It would be nice to use an `Mp_present` module for aliases to functor arguments and expand the ascription for substitution, but we don't have access to the target module type at substitution time (and indeed, we'd need an environment to do so). Consider for example
```ocaml
module type Inner_arg = sig
  type t
  val mk : unit -> t
end
module M : Inner_arg = ...
module type Arg = sig
    module F : functor (B : Inner_arg) -> sig
      module type S = module type of B end)
    end
    module F' : functor (B : Inner_arg) -> F(B).S
end
module Make (A : Arg) (B : Inner_arg) : A.F(B).S = struct
  include B
end
module F (A : Arg) = struct
  module F' = A.F'
  module N = F'(Make(A)(M))
end
```
Here, we would need the external environment (for `M`, `Make`), the functor environment (for `A`), and the structure item's local environment (for `F'`). To get the module type, we either have to construct this environment as we go in subst, pass some function that is able to infer this environment on the fly into subst, or perform the ascription when we initially type the functor. My preference is to go with the 3rd option, it being far simpler to implement. Then, we gain no advantage from making other aliases present if they otherwise need not be.

lpw25 commented:
> I've been playing with an approach that makes [...]

If I understand correctly, this is roughly equivalent to what I was proposing, except for:

> one of the above modules constrained by a named module type (module M : S = ...) use a variant of the named type with the aliases made absent

This will work, but it requires doing a deep traversal down `S` for every occurrence of `module M : S` to construct a new type with the absence information changed. I think that the benefit of doing this, in terms of linking fewer modules, will be quite small, whilst the cost, in terms of increased compilation times and worse error messages, will be quite high. (At Jane St. we find that signatures are frequently much larger than people expect -- think 10,000s of sub-components rather than 10s -- so we're trying to remove deep traversals of module types wherever possible.)

> To get the module type, we either have to construct this environment as we go in subst, pass some function that is able to infer this environment on the fly into subst, or perform the ascription when we initially type the functor. My preference is to go with the 3rd option, it being far simpler to implement. 

You don't have to do anything clever in `Subst`. You just add an additional function:
```ocaml
val add_ascribed_module: Ident.t -> Path.t -> module_type -> t -> t
```
where `add_ascribed_module id p mty` adds a substitution from `id` to `(p <: mty)`. This function is trivial to implement: in functor applications, projections and module aliases with existing ascriptions you just replace `id` by `p` same as usual, and in module aliases without ascriptions you just add `mty` as an ascription.

Then, when calculating the type of `Foo(Bar)` where `Foo` has type `functor (X : S) -> T` you use `add_ascribed_module X Bar S` where currently it uses `add_module X Bar`. By my count there are 4 such places and they all have `S` immediately available. This is because you must already know `S` when calculating the type of `Foo(Bar)` because you need to check `Bar` against it.

gasche commented:
We discussed this PR during a maintainer meeting today. Here is what I understood of the consensus.

1. There is wide agreement that this is a feature we want.
2. Leo (@lpw25) has an informal specification in mind, and he explained it in the thread above, and it's not clear your current implementation does that. The first step would be to align the PR with his suggestion.
3. Other maintainers would like to see a more formal specification; we asked @Drup to write one down, based on his existing document on modules (that does other, more complex things), to (a) check that it align's with @lpw25's specification, and (b) get others maintainers to read/approve it.
4. Independently of (3), @lpw25 and @Drup agreed to review the implementation of this PR once (2) is met.
5. If the specification is approved (3), and the implementation is approved (4), the PR will be merged.

gasche commented:
We also discussed the present/absent business, I'll report in #10435. We hope to tackle this easier aspect first, which should simplify the current PR.

mrmr1993 commented:
> You don't have to do anything clever in `Subst`. You just add an additional function:
> 
> ```ocaml
> val add_ascribed_module: Ident.t -> Path.t -> module_type -> t -> t
> ```
> 
> where `add_ascribed_module id p mty` adds a substitution from `id` to `(p <: mty)`. This function is trivial to implement: in functor applications, projections and module aliases with existing ascriptions you just replace `id` by `p` same as usual, and in module aliases without ascriptions you just add `mty` as an ascription.

@lpw25 to revisit this: your suggestion only works where the target is the substituted module. Consider instead
```ocaml
module type S = sig
  module M : sig val f : unit -> unit end
end
module F(A : S) = struct
  module M = A.M
end
module N = struct
  module M = struct
    let x = 1
    let f () = ()
  end
end
module Res = F(N)
```
In this scenario, we would apply the substitution `Subst.add_ascribed_module A N S subst`, so we need to do something to discover that `A.M = (N.M <: sig val f : unit -> unit)`.

lpw25 commented:
Sorry, I was aware of this issue but forgot to mention it. You need to change the type of the aliases from `Path.t * module_type option` to something like:
```ocaml
type alias_path =
  | Ident of Ident.t
  | Ascription of Path.t * module_type
  | Dot of alias_path * string
  | Apply of alias_path * Path.t
```
That way you can always replace an `Ident` with an `Ascription` during a substitution. Note that the `Ident.t`s inside the `Path.t`s do not need to have the ascription added.

Then you can write a function like:
```ocaml
val normalize_alias_path : Env.t -> alias_path -> Path.t * module_type option
```
and use that to get your current representation when you have an environment available.
